### PR TITLE
scheduler: resolve lock by batches.

### PR DESCRIPTION
**Pull request by disksing**
Resolve locks by batches to avoid generating a too huge write raft command.

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/storage/txn/scheduler.rs [hunk 1]
```diff
@@ -422,6 +417,30 @@ fn process_write_impl(cid: u64,
...
+                        try!(txn.commit(&k, ts));
+                    }
+                }
+                None => {
+                    for k in keys {
+                        try!(txn.rollback(&k));
+                    }
+                }
+            }
+            let pr = ProcessResult::NextCommand {
```
Comment (siddontang) [inline src/storage/txn/scheduler.rs]: Will the next rolled back seek the locks from the beginning? If yes, we may skip many tombstone keys.

**Comment by disksing** [comment]
PTAL @zhangjinpeng1987 @siddontang

File: src/storage/mvcc/reader.rs [hunk 2]
```diff
@@ -295,22 +295,38 @@ impl<'a> MvccReader<'a> {
...
         let mut locks = vec![];
         while cursor.valid() {
             let key = Key::from_encoded(cursor.key().to_vec());
             let lock = try!(Lock::parse(cursor.value()));
             if filter(&lock) {
-                locks.push((key, lock));
+                locks.push((key.clone(), lock));
+                if let Some(limit) = limit {
+                    if locks.len() >= limit {
+                        return Ok((locks, Some(key.append_ts(0))));
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/reader.rs]: why `append_ts(0)`

File: tests/storage/test_storage.rs [hunk 3]
```diff
@@ -467,6 +467,34 @@ fn test_txn_store_resolve_lock() {
...
 
+fn test_txn_store_resolve_lock_batch(n: usize) {
+    let store = new_assertion_storage();
+    for i in 0..n {
+        let key = format!("k{}", i);
+        store.prewrite_ok(vec![Mutation::Put((make_key(&key.as_bytes()), b"v".to_vec()))],
+                          b"k1",
+                          5);
+    }
+    store.resolve_lock_ok(5, Some(10));
```
Comment (zhangjinpeng1987) [inline tests/storage/test_storage.rs]: How about rollback?

File: src/storage/mvcc/reader.rs [hunk 4]
```diff
@@ -295,22 +295,38 @@ impl<'a> MvccReader<'a> {
...
+                        filter: F,
+                        limit: Option<usize>)
+                        -> Result<(Vec<(Key, Lock)>, Option<Key>)>
         where F: Fn(&Lock) -> bool
     {
         try!(self.create_lock_cursor());
         let mut cursor = self.lock_cursor.as_mut().unwrap();
-        cursor.seek_to_first();
+        let ok = match start {
+            Some(ref x) => try!(cursor.seek(x)),
```
Comment (siddontang) [inline src/storage/mvcc/reader.rs]: do we have test to cover `start` key?

**Comment by siddontang** [comment]
Please add a description for your PR.

File: src/storage/mvcc/reader.rs (hunk 4 above)
Comment (disksing) [inline src/storage/mvcc/reader.rs]: https://github.com/pingcap/tikv/pull/1271/files#diff-a581b51c6f8b34d75b3ce3e7c8037894R494
This test uses different count of locks to test ResolveLock, different values of `start` should be covered.

**Comment by disksing** [comment]
PTAL @siddontang @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**